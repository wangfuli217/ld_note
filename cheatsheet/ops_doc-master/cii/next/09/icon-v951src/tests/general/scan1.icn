#SRC: JCON

procedure main()
   local skips, vowels, uppers

   write(any('ab', "1234ab", 0, 7) | "fail")
   write(any('ab', "1234ab", 7, 0) | "fail")
   write(any('ab', "1234ab", 6, 0) | "fail")
   write(any('ab', "1234ab", 6) | "fail")
   write(any('ab', "1234ab") | "fail")
   "1234ab" ? write(any('ab', "1234ab") | "fail")
   "1234ab" ? write(any('ab') | "fail")
   "1234ab" ? { &pos := 6 & write(any('ab') | "fail") }
   write()

   write(many('ab', "ababac") | "fail")
   write(many('ab', "ababab") | "fail")
   write(many('ab', "cababab") | "fail")
   write()

   write(match("ab", "ababab") | "fail")
   write(match("ab", "bbabab") | "fail")
   write(match("ab", "ab") | "fail")
   write(match("ab", "a") | "fail")
   write()

   every write(find("ab", "ababab"|"ababa"|"ab"|"a"|"") | "done")
   write()

   every write(upto('ab', "abccab"|"cbabc"|"ab"|"a"|"") | "done")
   write()

   "123456" ? write(move(1), " ", move(2))
   "123456" ? { write(move(1)); write(move(2)); write(move(3));
		write(move(4)); write(move(-4))}
   write()

   "123456" ? write(tab(1), " ", tab(2))
   "123456" ? { write(tab(2)); write(tab(4)); write(tab(-1)); write(tab(0)); }
   "123456" ? { write(tab(4)); write(tab(1)); write(tab(3)); }
   write()

   every write("-35" ? bal('-'))
   every write("((2*x)+3)+(5*y)" ? bal('+'))
   every write("[+,[2,3]],[*,[5,10]]" ? bal(',', '[', ']'))
   write()

   "12345" ? { write(pos(1)); write(pos(-5)); write(pos(3)|"fail"); }
   write()

   "12345" ? { write(="123" | "fail") }
   "12345" ? { tab(2); write(="123" | "fail") }
   "12345" ? { tab(2); write(="23" | "fail") }

   # test upto, including some chars with 8th bit set
   write()
   skips := '\x00\x0f\x1e-<KZix\x87\x96\xa5\xb4\xc3\xd2\xe1\xf0\xff'
   vowels := 'aeoiuAEIOU'
   uppers := &cset -- &ascii
   &ascii ? {
      every writes(" ", "ascii?skips" | upto(skips) | "\n")
      every writes(" ", "ascii?vowls" | upto(vowels) | "\n")
      }
   &letters ? {
      every writes(" ", "letts?skips" | upto(skips) | "\n")
      every writes(" ", "letts?vowls" | upto(vowels) | "\n")
      }
   vowels ? {
      every writes(" ", "vowls?skips" | upto(skips) | "\n")
      every writes(" ", "vowls?letts" | upto(&letters) | "\n")
      every writes(" ", "vowls?ascii" | upto(&ascii) | "\n")
      }
   skips ? {
      every writes(" ", "skips?vowls" | upto(vowels) | "\n")
      every writes(" ", "skips?letts" | upto(&letters) | "\n")
      every writes(" ", "skips?ascii" | upto(&ascii) | "\n")
      every writes(" ", "skips?upprs" | upto(uppers) | "\n")
      every writes(" ", "skips?skips" | upto(skips) | "\n")
      }
   uppers ? {
      every writes(" ", "upprs?ascii" | upto(&ascii) | "\n")
      every writes(" ", "upprs?skips" | upto(skips) | "\n")
      }
end
