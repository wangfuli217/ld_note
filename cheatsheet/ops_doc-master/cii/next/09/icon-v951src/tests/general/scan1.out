fail
fail
7
7
fail
fail
fail
7

6
7
fail

3
fail
3
fail

1
3
5
1
3
1
done

1
2
5
6
2
3
4
1
2
1
done

1 23
1
23
456
3456

 1
1
23
45
6
123
123
12

1
10
10

1
1
fail

123
fail
23

 ascii?skips 1 16 31 46 61 76 91 106 121 
 ascii?vowls 66 70 74 80 86 98 102 106 112 118 
 letts?skips 11 26 35 50 
 letts?vowls 1 5 9 15 21 27 31 35 41 47 
 vowls?skips 8 
 vowls?letts 1 2 3 4 5 6 7 8 9 10 
 vowls?ascii 1 2 3 4 5 6 7 8 9 10 
 skips?vowls 8 
 skips?letts 6 7 8 9 
 skips?ascii 1 2 3 4 5 6 7 8 9 
 skips?upprs 10 11 12 13 14 15 16 17 18 
 skips?skips 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 
 upprs?ascii 
 upprs?skips 8 23 38 53 68 83 98 113 128 
